[profiler] add a function to allow adding preset user-defined metadata to traces

shengbao-zheng opened the pull request:
Summary:
`add_metadata_json` function in profiler can only work when being called during trace collection. However, sometimes we want to pass in some user-defined metadata and amend to the trace before trace collection starts, e.g. when the profiler is defined.
This PR add a function `preset_metadata_json` for this purpose. The preset metadata will be stored and amended to the trace later.

Differential Revision: D54678790

aaronenyeshi commented on torch/profiler/profiler.py:
```diff
@@ -262,6 +268,14 @@ def add_metadata_json(self, key: str, value: str):
 
+    def preset_metadata_json(self, key: str, value: str):
+        """
+        Preset a a user defined metadata when the profiler is not started
```
```suggestion
        Preset a user defined metadata when the profiler is not started
```

aaronenyeshi reviewed (approved):
LGTM, please check CI failures

shengbao-zheng committed:
[profiler] add a function to allow adding preset user-defined metadata to traces (#121487)

Summary:

`add_metadata_json` function in profiler can only work when being called during trace collection. However, sometimes we want to pass in some user-defined metadata and amend to the trace before trace collection starts, e.g. when the profiler is defined.
This PR add a function `preset_metadata_json` for this purpose. The preset metadata will be stored and amended to the trace later.

Reviewed By: aaronenyeshi

Differential Revision: D54678790
